hedera-node/docs/design/services/smart-contract-service/frictionless-airdrops-system-contracts.md

```diff
@@ -0,0 +1,165 @@
+
+## Implementation
+
+### New Solidity Structures
+
+We will introduce the following new structures to support the new functionality:
+
+`PendingAirdrop` - A struct that represents a pending airdrop request.
+
+```solidity
+struct PendingAirdrop {
+    address sender;
+    address receiver;
+
+    address token;
+    bool isNft;
+    int64 serial;
+}
+```
+
+`PendingAirdropRecord` - A struct that represents a pending airdrop record.
+
+```solidity
+struct PendingAirdropRecord {
+    PendingAirdrop pendingAirdrop;
+    uint64 amount;
+}
+```
+
+`NftId` - A struct that represents the Nft serials to be rejected.
+
+```solidity
+struct NftId {
+    address nft;
+    int64[] serials;
+}
+```
+
+### New Solidity Functions
+
+New system contract functions must be added to the `IHederaTokenService` interface to support airdropping tokens.
+
+| Function Selector Hash |                                                         Function Signature                                                          |                 Response                 |                                                                           |
+|------------------------|-------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------|---------------------------------------------------------------------------|
+| `0x2f348119`           | `function airdropTokens(TokenTransferList[] memory tokenTransfers) external returns (int64 responseCode)`                           | `ResponseCode`                           | The response code from the call                                           |
+| `0xc1acfe5d`           | `function cancelAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`                            | `ResponseCode`                           | The response code from the call                                           |
+| `0x241b6af7`           | `function claimAirdrops(PendingAirdrop[] memory pendingAirdrops) external returns (int64 responseCode)`                             | `ResponseCode`                           | The response code from the call                                           |
+| `0x012ea0b1`           | `function rejectTokens(address[] memory ftAddresses,  NftId[] memory nftIds) external returns (int64 responseCode)`                 | `ResponseCode`                           | The response code from the call                                           |
+| `0xc60ffe1b`           | `function getPendingAirdrops(address account) external returns (int64 responseCode, PendingAirdropRecord[] memory pendingAirdrops)` | `ResponseCode`, `PendingAirdropRecord[]` | The response code from the call and the pending airdrops for given sender |
```

Comment:
I don't think we should be adding this function in its form.
How would we handle when the number of airdrops is significant? We can' sent it all and if you send a partial amount how does a user get this information

Reply:
I see your point. This one is a proposal, as we only have the Mirror node to query for the Pending Airdrops where we have pagination. Maybe we should skip the query part and leave it to the Mirror node for now?

Reply:
I wonder what the use case is for a smart contract to call this (as opposed to a wallet or something else external to the blockchain) calling the mirror node)?